plone/app/uuid/utils.py

```diff
@@ -1,18 +1,56 @@
 # -*- coding: utf-8 -*-
+from AccessControl import Unauthorized
 from Products.CMFCore.utils import getToolByName
+from Products.ZCatalog.query import IndexQuery
 from zope.component.hooks import getSite
+from zope.globalrequest import getRequest
+
+
+def _catalog():
+    """Get and store portal_catalog on request.
+
+    This avoids looking up the site and the portal_catalog
+    each time one of the below functions is called.
+    """
+    request = getRequest()
+    try:
+        return request._catalog
+    except AttributeError:
```

Comment:
AFAIK (correct me if I am wrong): A traceback is expensive. getattr + condition is way cheaper.

Reply:
I think trying and catching an exception is usually cheaper. Let's test it.  I tried getting the catalog one million times. With the above exception handling: 0.53 seconds.
With gettattr: at first 0.56 seconds. On second try 0.52.  So seems about the same.
This was the code:

```
    catalog = getattr(request, '_catalog', _marker)
    if catalog is not _marker:
        return catalog
```